037-prototypes/proposal.md

```diff
@@ -0,0 +1,473 @@
+  the 'resources' concept and terminology. This RFC generalizes the design
+  further, switches to prototype-based terminology and allows prototypes to
+  handle arbitrary messages.
+
+* [RFC #1][rfc-1], now defunct, is similar to this proposal but had a concept
+  of "spaces" baked into the interface. This concept has been decoupled from
+  the resource interface and will be re-introduced in a later RFC that is
+  compatible with v1 and v2 resources.
+  * **Recommended reading**: [this comment][rfc-1-comment] outlines the thought
+    process that led to this RFC.
+
+* [concourse/concourse#534](https://github.com/concourse/concourse/issues/534)
+  was the first 'new resource interface' proposal which pre-dated the RFC
+  process.
+
+
+## Motivation
+
+* Support for 'reusable tasks' in order to stop the shoehorning of
+  functionality into the resource interface:
+  [concourse/rfcs#7](https://github.com/concourse/rfcs/issues/7).
+
+* Support for creating multiple versions from `put`:
+  [concourse/concourse#2660](https://github.com/concourse/concourse/issues/2660)
+
+* Support for deleting versions:
+  [concourse/concourse#362](https://github.com/concourse/concourse/issues/362),
+  [concourse/concourse#524](https://github.com/concourse/concourse/issues/524)
+
+* Having resource metadata immediately available via check:
+  [concourse/git-resource#193](https://github.com/concourse/git-resource/issues/193),
+  [concourse/concourse#1714](https://github.com/concourse/concourse/issues/1714)
+
+* Unifying `source` and `params` as just `config` so that resources don't have
+  to care where configuration is being set in pipelines:
+  [concourse/git-resource#172](https://github.com/concourse/git-resource/pull/172),
+  [concourse/bosh-deployment-resource#13](https://github.com/concourse/bosh-deployment-resource/issues/13),
+  [concourse/bosh-deployment-resource#6](https://github.com/concourse/bosh-deployment-resource/issues/6),
+  [concourse/cf-resource#20](https://github.com/concourse/cf-resource/pull/20),
+  [concourse/cf-resource#25](https://github.com/concourse/cf-resource/pull/25),
+  [concourse/git-resource#210](https://github.com/concourse/git-resource/pull/210)
+
+* Make resource actions reentrant so that we no longer receive `unexpected EOF`
+  errors when reattaching to an in-flight build whose resource action completed
+  while we weren't attached:
+  [concourse/concourse#1580](https://github.com/concourse/concourse/issues/1580)
+
+* Support for showing icons for resources in the web UI:
+  [concourse/concourse#788](https://github.com/concourse/concourse/issues/788),
+  [concourse/concourse#3220](https://github.com/concourse/concourse/pull/3220),
+  [concourse/concourse#3581](https://github.com/concourse/concourse/pull/3581)
+
+* Standardize TLS configuration so every resource doesn't implement their own
+  way: [concourse/rfcs#9](https://github.com/concourse/rfcs/issues/9)
```

Comment:
Some discussion of this in https://github.com/concourse/concourse/issues/5190, for reference.